core/block.js

```diff
@@ -814,10 +814,10 @@ Block.prototype.setParent = function(newParent) {
  * @return {!Array<!Block>} Flattened array of blocks.
  */
 Block.prototype.getDescendants = function(ordered) {
-  const blocks = [this];
+  let blocks = [this];
   const childBlocks = this.getChildren(ordered);
   for (let child, i = 0; (child = childBlocks[i]); i++) {
-    blocks.push.apply(blocks, child.getDescendants(ordered));
+    blocks = [...blocks, ...child.getDescendants(ordered)];
```

Comment:
Can you check the performance delta of this?  Especially after being cross-compiled.  Usually (but not always) ES6 stuff is slower, and this looks like a critical area.  There's another similar change later in this PR too.

Reply:
I suspect that this version will be slower simply because (unless the JIT is very clever) it creates a new object and copies all the entries over instead of appending to the existing one.

An alternative that might be both more readable and performant than the original is `blocks.push(blocks, ...child.getDescendants(ordered))`.

Reply:
Good call. 

TLDR: `blocks.push(...child.getDescendants(ordered)` is the move here.

Note that there's no need to use `push.apply` given that both objects are already arrays.

### Details

I benchmarked two ways: using [neil's blog](https://neil.fraser.name/news/2021/06/07/) and using [JSbench.me](https://jsbench.me/).

For each benchmark, I had the following setup code:
```
let array1 = [];
const array2 = [];
for (let i = 0; i < 1000; i++) {
  array1.push(i);
  array2.push(i + 2000);
}
```

#### Test cases

Then I tested three cases:
Spread only:
```
for (let j = 0; j < 1000; j++) {
  array1 = [...array1, ...array2];
}
```

Push + spread:
```
for (let j = 0; j < 1000; j++) {
  array1.push(...array2);
}
```

Concat:
```
for (let j = 0; j < 1000; j++) {
  array1 = array1.concat(array2);
}
```

Finally, I tried this with no looping in the actual test cases--just concatenating once.

#### Results

With both benchmarks, and for both looping and not, `[...arr1, ...arr2]` is at least an order of magnitude slower. `arr1.push(...arr2)` is faster than `arr1.concat(arr2)`, although the two sites give wildly different numbers for concat.

### Does it matter?
For this case: maybe. For the `renderList` case below: yes, absolutely, because that's a very hot path.

Reply:
Fixed.

Reply:
Also relevant, Babel transpiles this code:

`  array1.push(...array2);`

into this code:

`  array1.push.apply(array1, array2);`

Which brings us back full circle to the original code.

However, that's only if the compiler can detect through static analysis that array2 is indeed an array.  If it can't figure this out, then it generates the following:

```
function _toConsumableArray(arr) {
  return (
    _arrayWithoutHoles(arr) ||
    _iterableToArray(arr) ||
    _unsupportedIterableToArray(arr) ||
    _nonIterableSpread()
  );
}

function _nonIterableSpread() {
  throw new TypeError(
    "Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."
  );
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n))
    return _arrayLikeToArray(o, minLen);
}

function _iterableToArray(iter) {
  if (
    (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null) ||
    iter["@@iterator"] != null
  )
    return Array.from(iter);
}

function _arrayWithoutHoles(arr) {
  if (Array.isArray(arr)) return _arrayLikeToArray(arr);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;
  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }
  return arr2;
}

array1.push.apply(array1, _toConsumableArray(array2));
```

Reply:
I've run `npm build` on this code, and each of the six instances of `...` get compiled to a variation of:

`b.push.apply(b,$jscomp.arrayFromIterable(d.getDescendants(a)))`

Same as Babel when it can't be certain that the argument is an array.

So, if this code is intended to be cross-compiled to ES5, then could be a problem since these look like functions which are performance-sensitive.

Reply:
Okay, I reverted it entirely and disabled that rule in `.eslintrc.json`

Reply:
Seems like we ought to be able to improve the typing info and thereby allow Closure Compiler to better optimise this—but it would probably require advanced optimisation.

Reply:
Update: unlike Babel, [`$jscomp.arrayFromIterable` fast-paths the case where its argument is an array](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/js/es6/util/arrayfromiterable.js), so there's a good chance that the performance of the compiled code (even absent advanced optimisations) will end up being basically the same as the `Array.prototype.push.apply` implementation once the JS engine has worked its magic—though it will vary depending on the competence of the JIT, of course.  No objection to continuing to use the straight-forward implementation here, but maybe a reminder to try to benchmark the actual code that we ship…